shared/subspace-data-retrieval/src/object_fetcher/partial_object.rs

```diff
@@ -0,0 +1,736 @@
+        self.prefix_data.append(&mut new_data.prefix_data);
+        // Keep the remaining non-padding and potential padding bytes
+        self.prefix_data.append(&mut new_data.maybe_padding_data);
+        self.suffix_data = new_data.suffix_data;
+
+        // There's only one data length, but now we have multiple padding lengths
+        let data_length = self.shortest_object_length();
+        // Avoid a range panic if there's actually no padding, by overwriting the existing length
+        for ignored_padding_length in 0..=object_max_segment_padding_length {
+            self.lengths.insert(ObjectLength {
+                data_length,
+                ignored_padding_length,
+            });
+        }
+    }
+
+    /// Add padding-less `piece_data` to the end of this partial object.
+    pub fn add_piece_data_without_padding(&mut self, mut piece_data: Vec<u8>) {
+        if self.has_padding() {
+            // If there might be padding, or there is suffix data, new data must be added at the end
+            // of the suffix data.
+            self.suffix_data.append(&mut piece_data);
+        } else {
+            // For performance, add the data to the existing prefix data.
+            self.prefix_data.append(&mut piece_data);
+        }
+    }
+
+    /// Returns true if this object has padding, or might have padding.
+    ///
+    /// Objects can have no padding in three different ways:
+    /// - they aren't at the end of a segment
+    /// - they are at the end of a segment, but all potential padding bytes are non-zero
+    /// - we've checked the non-zero ignored padding lengths, and they were all invalid
+    ///
+    /// Panics if the object has no valid lengths left.
+    fn has_padding(&self) -> bool {
+        !self.suffix_data.is_empty()
+            || self.lengths.len() > 1
+            || self.largest_ignored_padding_length() > 0
+    }
+
+    /// Returns the longest padding length the object could possibly have.
+    /// Adding padding can change the object length, but this is always returns the longest
+    /// possible amount of padding bytes.
+    ///
+    /// Panics if the object has no valid lengths left.
+    fn largest_ignored_padding_length(&self) -> usize {
+        self.lengths
+            .iter()
+            .map(|length| length.ignored_padding_length)
+            .max()
+            .expect("other methods return an error if lengths becomes empty; qed")
+    }
+
+    /// Check the hash against all possible objects with enough data.
+    /// If a possible object's data doesn't match the mapping hash, its length is removed.
+    ///
+    /// Returns valid object data, `Ok(None)` if we need more data, or an error if all possible
+    /// objects have invalid hashes.
+    ///
+    /// Panics if it is called after there are no possible lengths left.
+    /// (The previous call returned an error, so the caller should have stopped checking.)
+    pub fn check_available_objects(
```

Comment:
this fn name is not very clear. I think something like `is_object_recontructed` or something similar would make more sense IMHO

Reply:
I went with `try_reconstruct_object()`, because it's an object reconstruction operation that can fail permanently (all possible object/padding lengths do not match the hash), or temporarily (needs more data downloaded).